src/main/scala/uk/gov/nationalarchives/omega/api/services/Dispatcher.scala

```diff
@@ -0,0 +1,107 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.gov.nationalarchives.omega.api.services
+
+import cats.data.Validated.{ Invalid, Valid }
+import cats.data.{ Validated, ValidatedNec }
+import cats.effect.IO
+import cats.effect.std.Queue
+import uk.gov.nationalarchives.omega.api.business._
+import uk.gov.nationalarchives.omega.api.business.echo.{ EchoRequest, EchoResponse, EchoService }
+import uk.gov.nationalarchives.omega.api.services.ServiceIdentifier.ECHO001
+
+import java.util.UUID
+
+class Dispatcher(val localProducer: LocalProducer) {
+
+  def run(dispatcherId: Int)(q: Queue[IO, LocalMessage]): IO[Unit] =
+    for {
+      requestMessage <- q.take
+      _ <- IO.delay {
+             println(s"Dispatcher # $dispatcherId, processing message id: ${requestMessage.persistentMessageId}")
+           }
+      serviceRequest                  <- createServiceRequest(requestMessage)
+      businessService                 <- IO.pure(serviceRequest._1)
+      businessServiceRequest          <- IO.pure(serviceRequest._2)
+      validatedBusinessServiceRequest <- validateBusinessServiceRequest(businessService, businessServiceRequest)
+      businessResult                  <- execBusinessService(businessService, validatedBusinessServiceRequest)
+      res                             <- sendResultToJmsQueue(businessResult, requestMessage)
+    } yield res
+
+  private def createServiceRequest(localMessage: LocalMessage): IO[(BusinessService, BusinessServiceRequest)] =
+    localMessage.serviceId match {
+      case ECHO001 =>
+        IO.delay {
+          Tuple2(
+            new EchoService(),
+            EchoRequest(localMessage.messageText)
+          )
+        }
+      // add more service IDs here
+    }
+
+  private def validateBusinessServiceRequest(
+    businessService: BusinessService,
+    businessServiceRequest: BusinessServiceRequest
+  ): IO[ValidatedNec[RequestValidationError, BusinessServiceRequest]] =
+    IO.delay {
+      businessService match {
+        case value: RequestValidation =>
+          value.validateRequest(businessServiceRequest)
+        case _ =>
+          Validated.valid(businessServiceRequest)
+      }
+    }
+
+  private def execBusinessService[T <: BusinessServiceRequest, U <: BusinessServiceResponse, E <: BusinessServiceError](
+    businessService: BusinessService,
+    validatedBusinessServiceRequest: ValidatedNec[RequestValidationError, T]
+  ): IO[ValidatedNec[RequestValidationError, Either[BusinessServiceError, BusinessServiceResponse]]] =
+    IO.delay {
```

Comment:
Why does this method use IO?

Reply:
This is where the business service gets executed. As this is could be time-intensive it seems reasonable to wrap this in IO.